orchestrator/migrations/versions/schema/2023-03-06_e05bb1967eff_add_subscriptions_search_view.py

```diff
@@ -0,0 +1,156 @@
+                          JOIN fixed_inputs fi ON p.product_id = fi.product_id
+                 GROUP BY s.subscription_id),
+     cust_info AS (SELECT s.subscription_id,
+                          string_agg('customer_description: ' || scd.description, ', ') AS cust_info
+                   FROM subscriptions s
+                            JOIN subscription_customer_descriptions scd ON s.subscription_id = scd.subscription_id
+                   GROUP BY s.subscription_id)
+SELECT s.subscription_id,
+       to_tsvector('simple',
+                   concat_ws(', ',
+                             spi.sub_info,
+                             spi.prod_info,
+                             fi.fi_info,
+                             rti.rt_info,
+                             ci.cust_info)
+           ) as tsv
+FROM subscriptions s
+         LEFT JOIN sub_prod_info spi ON s.subscription_id = spi.subscription_id
+         LEFT JOIN fi_info fi ON s.subscription_id = fi.subscription_id
+         LEFT JOIN rt_info rti ON s.subscription_id = rti.subscription_id
+         LEFT JOIN cust_info ci ON s.subscription_id = ci.subscription_id
+         """
+    subscriptions_search_view_ddl = (
+        f"CREATE MATERIALIZED VIEW IF NOT EXISTS subscriptions_search AS {subscriptions_search_query}"
+    )
+
+    refresh_subscriptions_search_fn = """
+      CREATE OR REPLACE FUNCTION refresh_subscriptions_search_view()
+  RETURNS TRIGGER LANGUAGE plpgsql
+  AS $$
+  BEGIN
+  REFRESH MATERIALIZED VIEW CONCURRENTLY subscriptions_search;
+  RETURN NULL;
+  END $$;
+"""
+
+    conn.execute(subscriptions_search_view_ddl)
+    conn.execute(refresh_subscriptions_search_fn)
+    conn.execute("CREATE INDEX subscriptions_search_tsv_idx ON subscriptions_search USING GIN (tsv);")
+
+    # Refresh the view when dependent tables change
+    conn.execute(
+        "CREATE CONSTRAINT TRIGGER fi_refresh_search AFTER UPDATE ON fixed_inputs DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE FUNCTION refresh_subscriptions_search_view();"
+    )
+    conn.execute(
+        "CREATE CONSTRAINT TRIGGER products_refresh_search AFTER UPDATE ON products DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE FUNCTION refresh_subscriptions_search_view();"
+    )
+    conn.execute(
+        "CREATE CONSTRAINT TRIGGER sub_cust_desc_refresh_search AFTER INSERT OR UPDATE OR DELETE ON subscription_customer_descriptions DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE FUNCTION refresh_subscriptions_search_view();"
+    )
+    conn.execute(
+        "CREATE CONSTRAINT TRIGGER siv_refresh_search AFTER INSERT OR UPDATE OR DELETE ON subscription_instance_values DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE FUNCTION refresh_subscriptions_search_view();"
+    )
+    conn.execute(
+        "CREATE CONSTRAINT TRIGGER sub_refresh_search AFTER INSERT OR UPDATE OR DELETE ON subscriptions DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE FUNCTION refresh_subscriptions_search_view();"
```

Comment:
How expensive is this refresh? I think this is worth testing on staging with a production export.

And: if a transaction is committed which modifies X rows, will the view be refreshed X times or just 1?

Reply:
The `DEFERRABLE INITIALLY DEFERRED` option makes it execute once at the end of a transaction. Would indeed be good to test on a large dataset

Reply:
Takes about one second on a database with ~7200 subscriptions, ~18000 subscription instances and ~65000 subscription instance values (on my machine)